### PR TITLE
fix(codebuild): allow passing the ARN of the Secret in environment variables

### LINKED ISSUE #12703: (codebuild): Secret manager env vars get the wrong arn on grant
When adding environment variables to codebuild Projects, cdk adds the grant read for the secret. However, when adding a secret using the `<secretName>:<secretKey>` format, the generated arn includes the `secretKey` causing the arn to be wrong (`aws::secretsmanager:eu-central-1:00000000000:secret:secretName:secretKey-??????`)

### Reproduction Steps

Create a codebuild Project using a secret env var with the `<secretName>:<secretKey>` format

### What did you expect to happen?

the arn on the policy not to include the secretKey

### What actually happened?

the arn on the policy included the secretKey

### Environment

  - **CDK CLI Version  : 1.86.0**
  - **Framework Version: 1.86.0**
  - **Node.js Version: 13.8.0**
  - **OS               : MacOS**
  - **Language (Version):TypeScript (3.9.7)**

### Other

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by skinny85**
In the SecretsManager-typed environment variables in CodeBuild,
the code in the Project class assumed those would be passed as names.
As it turns out, CodeBuild also allows passing there entire ARNs of secrets
(both partial, and full), and also optional qualifiers,
separated by colons, that specify SecretsManager attributes like the JSON key,
or the secret version.
Add handling of all of these cases.

Fixes #12703

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 1]
```diff
@@ -749,27 +751,58 @@ export class Project extends ProjectBase {
...
         // save SecretsManager env variables
         if (envVariable.type === BuildEnvironmentVariableType.SECRETS_MANAGER) {
-          secretsManagerSecrets.push(Stack.of(principal).formatArn({
-            service: 'secretsmanager',
-            resource: 'secret',
-            // we don't know the exact ARN of the Secret just from its name, but we can get close
-            resourceName: `${envVariable.value}-??????`,
-            sep: ':',
-          }));
+          if (Token.isUnresolved(envVariableValue)) {
```
Comment (njlynch) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts:754]: Is there a test for this condition? I'm not seeing it.

Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts:754]: Yep - those are the `can be provided as the ARN attribute of a new Secret` and `can be provided as the ARN attribute of a new Secret, followed by a JSON key` tests.

Comment (njlynch) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts:754]: Ah, I guess I was looking for a test where a `Fn:Split` call was expected, as we might be splitting a token. Now I see that we're assuming the input is _partially_ an unresolved token (e.g., `${secretArn}:jsonKey`).

If I understand correctly, you're supporting something like the above, but not a "full" token'ed value. So if I create a custom resource which returns back `my-secret-name:jsonKey` as a Token, then we will be passing the Token as a whole to `secretsManagerIamResources`, and the grant won't work. Am I reading that correctly now?

Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts:754]: Yep, you're reading it correctly.

What's the behavior that you would want to see here?

Comment (njlynch) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts:754]: Frankly, I'm not sure. I don't know that there's a good solution here. I'm hoping this is a relatively edge-of-the-corner case. :D

**Review by njlynch — APPROVED** [review]

**Pull request merged**